src/evaluators/BinaryExpression.js

```diff
@@ -40,38 +40,40 @@ let unknownValueOfOrToString = "might be an object with an unknown valueOf or to
 export function getPureBinaryOperationResultType(
   realm: Realm, op: BabelBinaryOperator, lval: Value, rval: Value, lloc: ?BabelNodeSourceLocation, rloc: ?BabelNodeSourceLocation
 ): void | typeof Value {
+  function reportErrorIfNotPure(purityTest: (Realm, Value) => boolean, typeIfPure: typeof Value): void | typeof Value {
+    let leftPure = purityTest(realm, lval);
+    let rightPure = purityTest(realm, rval);
+    if (leftPure && rightPure) return BooleanValue;
```

Comment:
And if so, where could we use an additional invariant, or what test case is missing, to detect that?